File: riistaandroid/src/main/java/fi/riista/mobile/storage/AnnouncementSync.java
package fi.riista.mobile.storage;


import android.util.Log;

import java.util.List;

import fi.riista.mobile.RiistaApplication;
import fi.riista.mobile.models.announcement.Announcement;
import fi.riista.mobile.network.ListAnnouncementsTask;

public class AnnouncementSync {

    public interface AnnouncementSyncListener {
        void onFinish();
    }

    public void sync(final AnnouncementSyncListener listener) {
        ListAnnouncementsTask task = new ListAnnouncementsTask(RiistaApplication.getInstance().getWorkContext()) {
            @Override
            protected void onFinishObjects(List<Announcement> results) {
                StorageDatabase.getInstance().updateAnnouncements(results, new StorageDatabase.UpdateListener() {
                    @Override
                    public void onUpdate() {
                        listener.onFinish();
                    }

                    @Override
                    public void onError() {
                        listener.onFinish();
                    }
                });
            }

            @Override
            protected void onError() {
                Log.d("Announcements", "Sync error: " + getError().getMessage());

                listener.onFinish();
            }
        };
        task.start();
    }
}

File: riistaandroid/src/main/java/fi/riista/mobile/sync/AnnouncementSync.kt
package fi.riista.mobile.sync

import android.util.Log
import fi.riista.mobile.di.DependencyQualifiers.APPLICATION_WORK_CONTEXT_NAME
import fi.riista.mobile.models.announcement.Announcement
import fi.riista.mobile.network.ListAnnouncementsTask
import fi.riista.mobile.storage.StorageDatabase
import fi.riista.mobile.storage.StorageDatabase.UpdateListener
import fi.vincit.androidutilslib.context.WorkContext
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Singleton
class AnnouncementSync @Inject constructor(
        @Named(APPLICATION_WORK_CONTEXT_NAME) private val syncWorkContext: WorkContext) {

    fun sync(onCompleteTask: () -> Unit) {
        val task: ListAnnouncementsTask = object : ListAnnouncementsTask(syncWorkContext) {
            override fun onFinishObjects(results: List<Announcement>) {
                StorageDatabase.getInstance()?.updateAnnouncements(results, object : UpdateListener {
                    override fun onUpdate() {
                        onCompleteTask()
                    }

                    override fun onError() {
                        onCompleteTask()
                    }
                })
            }

            override fun onError() {
                Log.d(TAG, "Sync error: ${error.message}")
                onCompleteTask()
            }
        }
        task.start()
    }

    companion object {
        private const val TAG = "AnnouncementsSync"
    }
}
